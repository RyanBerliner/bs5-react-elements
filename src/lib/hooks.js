import {useEffect, useRef} from 'react';

/**
 * Hook to initialize a component, bind events, and expose bootstrap apis.
 *
 * @param {Object} Component Bootstrap component object (Modal, Tooltip, etc)
 * @param {Object} config (optional) Configuration object of the component
 * @param {Object} componentRef (optional) A ref to reference the component
 * @param {Object} domRef A ref pointing to the dom element of the component
 * @param {Map} events Mapping of bootstrap event names and function handlers
 */
export function useBootstrap(
    Component,
    config,
    componentRef,
    domRef,
    events,
) {
  const initialConfig = useRef(config);

  useEffect(() => {
    const component = Component.getInstance(domRef.current) ||
      new Component(domRef.current, initialConfig.current);
    if (componentRef) {
      componentRef.current = component;
    }

    return () => component.dispose();
  }, [Component, componentRef, domRef]);

  useEffect(() => {
    const el = domRef.current;
    events.forEach((value, key) => {
      el.addEventListener(key, value);
    });

    return () => {
      events.forEach((value, key) => {
        el.removeEventListener(key, value);
      });
    };
  }, [domRef, events]);
}
