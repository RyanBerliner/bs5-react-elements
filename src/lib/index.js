export {Alert} from './Alert.jsx';
export {Carousel} from './Carousel.jsx';
export {Dropdown} from './Dropdown.jsx';
export {Modal} from './Modal.jsx';
export {Offcanvas} from './Offcanvas.jsx';
export {Popover} from './Popover.jsx';
export {Tab} from './Tab.jsx';
export {Toast} from './Toast.jsx';
export {Tooltip} from './Tooltip.jsx';
