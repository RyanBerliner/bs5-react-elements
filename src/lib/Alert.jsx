import React, {useRef, useMemo, useCallback} from 'react';
import {unmountComponentAtNode} from 'react-dom';
import PropTypes from 'prop-types';
import {Alert} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.0/components/alerts/">Bootstrap alert component.</a>
 */
function AlertComponent({
  onClosed,
  onClose,
  container,
  component,
  children,
  ...props
}) {
  const componentElement = useRef();

  const _onClosed = useCallback((event) => {
    unmountComponentAtNode(container.current);

    if (onClosed) {
      onClosed(event);
    }
  }, [onClosed]);

  const events = useMemo(() => {
    return new Map([
      ['closed.bs.alert', _onClosed],
      ['close.bs.alert', onClose],
    ]);
  }, [_onClosed, onClose]);

  useBootstrap(Alert, undefined, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

AlertComponent.propTypes = {
  /**
   * Handler for the closed.bs.alert event
   */
  onClosed: PropTypes.func,
  /**
   * Handler for the close.bs.alert event
   */
  onClose: PropTypes.func,
  /**
   * React ref refering to the container the element is wrapped in
   */
  container: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]).isRequired(),
  /**
   * React ref that will be assigned the component instance
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Alert contents
   */
  children: PropTypes.node,
};

AlertComponent.displayName = 'Alert';

export {AlertComponent as Alert};
