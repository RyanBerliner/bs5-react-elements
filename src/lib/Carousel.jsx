import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Carousel} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.2/components/carousel/">Bootstrap carousel component.</a>
 */
function CarouselComponent({
  onSlid,
  onSlide,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['slid.bs.carousel', onSlid],
      ['slide.bs.carousel', onSlide],
    ]);
  }, [onSlid, onSlide]);

  useBootstrap(Carousel, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

CarouselComponent.propTypes = {
  /**
   * Handler for the slid.bs.carousel event
   * @since 0.2.0
   */
  onSlid: PropTypes.func,
  /**
   * Handler for the slide.bs.carousel event
   * @since 0.2.0
   */
  onSlide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.2.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.2.0
   */
  config: PropTypes.object,
  /**
   * Dropdown contents
   * @since 0.2.0
   */
  children: PropTypes.node,
};

CarouselComponent.displayName = 'Carousel';

export {CarouselComponent as Carousel};
