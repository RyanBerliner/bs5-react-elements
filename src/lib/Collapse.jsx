import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Collapse} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.0/components/collapse/">Bootstrap collapse component.</a>
 */
function CollapseComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.collapse', onShown],
      ['show.bs.collapse', onShow],
      ['hidden.bs.collapse', onHidden],
      ['hide.bs.collapse', onHide],
    ]);
  }, [onShown, onShow, onHidden, onHide]);

  useBootstrap(Collapse, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

CollapseComponent.propTypes = {
  /**
   * Handler for the shown.bs.collapse event
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.collapse event
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.collapse event
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.collapse event
   */
  onHide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   */
  config: PropTypes.object,
  /**
   * Collapse contents
   */
  children: PropTypes.node,
};

CollapseComponent.displayName = 'Collapse';

export {CollapseComponent as Collapse};
