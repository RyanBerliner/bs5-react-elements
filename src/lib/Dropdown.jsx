import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Dropdown} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.3/components/dropdowns/">Bootstrap dropdown component.</a>
 */
function DropdownComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  component,
  config,
  children,
  as: ElementType,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.dropdown', onShown],
      ['show.bs.dropdown', onShow],
      ['hidden.bs.dropdown', onHidden],
      ['hide.bs.dropdown', onHide],
    ]);
  }, [onShown, onShow, onHidden, onHide]);

  useBootstrap(Dropdown, config, component, componentElement, events);

  return (
    <ElementType ref={componentElement} {...props}>
      {children}
    </ElementType>
  );
}

DropdownComponent.propTypes = {
  /**
   * Handler for the shown.bs.dropdown event
   * @since 0.2.0
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.dropdown event
   * @since 0.2.0
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.dropdown event
   * @since 0.2.0
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.dropdown event
   * @since 0.2.0
   */
  onHide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.2.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.2.0
   */
  config: PropTypes.object,
  /**
   * Dropdown contents
   * @since 0.2.0
   */
  children: PropTypes.node,
  /**
   * The element type of the dropdown
   * @since 0.2.0
   */
  as: PropTypes.elementType,
};

DropdownComponent.defaultProps = {
  as: 'button',
};

DropdownComponent.displayName = 'Dropdown';

export {DropdownComponent as Dropdown};
