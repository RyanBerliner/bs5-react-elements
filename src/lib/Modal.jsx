import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Modal} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.3/components/modal/">Bootstrap modal component.</a>
 */
function ModalComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  onHidePrevented,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.modal', onShown],
      ['show.bs.modal', onShow],
      ['hidden.bs.modal', onHidden],
      ['hide.bs.modal', onHide],
      ['hidePrevented.bs.modal', onHidePrevented],
    ]);
  }, [onShown, onShow, onHidden, onHide, onHidePrevented]);

  useBootstrap(Modal, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

ModalComponent.propTypes = {
  /**
   * Handler for the shown.bs.modal event
   * @since 0.1.0
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.modal event
   * @since 0.1.0
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.modal event
   * @since 0.1.0
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.modal event
   * @since 0.1.0
   */
  onHide: PropTypes.func,
  /**
   * Handler for the hidePrevented.bs.modal event
   * @since 0.1.0
   */
  onHidePrevented: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.1.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.1.0
   */
  config: PropTypes.object,
  /**
   * Modal contents
   * @since 0.1.0
   */
  children: PropTypes.node,
};

ModalComponent.displayName = 'Modal';

export {ModalComponent as Modal};
