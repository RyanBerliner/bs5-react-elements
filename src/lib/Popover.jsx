import React, {useRef, useMemo, useCallback, useState} from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import {Popover} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.1/components/popovers/">Bootstrap popover component.</a>
 */
function PopoverComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  onInserted,
  component,
  config,
  children,
  as: ElementType,
  renderTitle,
  renderContent,
  ...props
}) {
  const componentElement = useRef();
  const [tip, setTip] = useState();

  const wrappedOnInserted = useCallback((event) => {
    const tip = Popover.getInstance(componentElement.current).getTipElement();

    if (renderTitle) {
      tip.querySelector('.popover-header').innerHTML = '';
    }

    if (renderContent) {
      tip.querySelector('.popover-body').innerHTML = '';
    }

    setTip(tip);

    if (onInserted) {
      onInserted(event);
    }
  }, [onInserted, renderTitle, renderContent]);

  const wrappedOnHide = useCallback((event) => {
    setTip(null);

    if (onHide) {
      onHide(event);
    }
  }, [onHide]);

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.popover', onShown],
      ['show.bs.popover', onShow],
      ['hidden.bs.popover', onHidden],
      ['hide.bs.popover', wrappedOnHide],
      ['inserted.bs.popover', wrappedOnInserted],
    ]);
  }, [onShown, onShow, onHidden, wrappedOnHide, wrappedOnInserted]);

  if (!config) {
    config = {};
  }

  if ((renderTitle || renderContent) && config.animation !== false) {
    config.animation = false;
  }

  useBootstrap(Popover, config, component, componentElement, events);

  if (renderTitle) {
    props.title = ' ';
  }

  if (renderContent) {
    props['data-bs-content'] = ' ';
  }

  return (
    <ElementType ref={componentElement} {...props}>
      {children}
      {(tip && renderTitle) && ReactDOM.createPortal(
          renderTitle(Popover.getInstance(componentElement.current)),
          tip.querySelector('.popover-header'),
      )}
      {(tip && renderContent) && ReactDOM.createPortal(
          renderContent(Popover.getInstance(componentElement.current)),
          tip.querySelector('.popover-body'),
      )}
    </ElementType>
  );
}

PopoverComponent.propTypes = {
  /**
   * Handler for the shown.bs.popover event
   * @since 0.2.0
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.popover event
   * @since 0.2.0
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.popover event
   * @since 0.2.0
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.popover event
   * @since 0.2.0
   */
  onHide: PropTypes.func,
  /**
   * Handler for the inserted.bs.popover event
   * @since 0.2.0
   */
  onInserted: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.2.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.2.0
   */
  config: PropTypes.object,
  /**
   * Popover contents
   * @since 0.2.0
   */
  children: PropTypes.node,
  /**
   * The element type of the popover
   * @since 0.2.0
   */
  as: PropTypes.elementType,
  /**
   * Render popover title as a react component
   * @since 2.0.0
   */
  renderTitle: PropTypes.func,
  /**
   * Render popover content as a react component
   * @since 2.0.0
   */
  renderContent: PropTypes.func,
};

PopoverComponent.defaultProps = {
  as: 'span',
};

PopoverComponent.displayName = 'Popover';

export {PopoverComponent as Popover};
