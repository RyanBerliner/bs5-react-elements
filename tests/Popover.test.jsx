import React, {useEffect, useRef, forwardRef} from 'react';
import PropTypes from 'prop-types';
import {Popover} from '../src/lib/Popover';
import {majorMinor} from './utils';
import {render, screen, fireEvent} from '@testing-library/react';

/**
 * A boilerplate popover.
 */
function TestPopover(props) {
  return <Popover
    title="Lorem ipsum"
    data-bs-content="Lorem ipsum dolor."
    data-bs-toggle="popover"
    {...props}
  >
    Test popover
  </Popover>;
}

/**
 * Test component for mounting in render props of popovers
 */
function TestComponent(props) {
  return <strong>Lorem ipsum</strong>;
}

/**
 * Clicks the test test popover.
 */
function clickPopover() {
  fireEvent.click(screen.getByText('Test popover'));
}

/**
 * Clicks the test popover after a small delay (literally minimal)
 */
function delayedClickPopover() {
  setTimeout(() =>{
    fireEvent.click(screen.getByText('Test popover'));
  });
}

// Timeouts are used to prevent disposing of bootstrap components that are in
// the middle of a transition. This is a bootstrap bug. TODO: Submit an issue.

describe('events bound to props', () => {
  test('show.bs.popover event bound to onShow prop', (done) => {
    /**
     * show.bs.popover event handler.
     *
     * @param {Object} event
     */
    function onShow(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestPopover onShow={onShow} />);
    clickPopover();
  });

  test('shown.bs.popover event bound to onShown prop', (done) => {
    /**
     * shown.bs.popover event handler.
     *
     * @param {Object} event
     */
    function onShown(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestPopover onShown={onShown} />);
    clickPopover();
  });

  test('hide.bs.popover event bound to onHide prop', (done) => {
    /**
     * hide.bs.popover event handler.
     *
     * @param {Object} event
     */
    function onHide(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestPopover onShown={clickPopover} onHide={onHide} />);
    clickPopover();
  });

  test('hidden.bs.popover event bound to onHidden prop', (done) => {
    /**
     * hidden.bs.popover event handler.
     *
     * @param {Object} event
     */
    function onHidden(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestPopover onShown={clickPopover} onHidden={onHidden} />);
    clickPopover();
  });

  test('inserted.bs.popover event bound to onInserted prop', (done) => {
    /**
     * inserted.bs.popover event handler.
     *
     * @param {Object} event
     */
    function onInserted(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestPopover onInserted={onInserted} />);
    clickPopover();
  });
});

describe('render title', () => {
  test('properly mounts react component in title', (done) => {
    /**
     * On inserted handler that verifys popover content
     */
    function onInserted() {
      setTimeout(() => {
        const popoverHeader = document.querySelector('.popover-header');
        expect(popoverHeader.innerHTML).toBe('<strong>Lorem ipsum</strong>');
        done();
      });
    }

    render(
        <TestPopover
          onInserted={onInserted}
          renderTitle={() => <TestComponent />}
        />,
    );
    clickPopover();
  });

  test('properly unmounts react component in popover', (done) => {
    render(
        <TestPopover
          onInserted={delayedClickPopover}
          onHidden={() => done()}
          renderTitle={() => <TestComponent />}
        />,
    );
    clickPopover();
  });
});

describe('render content', () => {
  test('properly mounts react component in content', (done) => {
    /**
     * On inserted handler that verifys popover content
     */
    function onInserted() {
      setTimeout(() => {
        const popoverBody = document.querySelector('.popover-body');
        expect(popoverBody.innerHTML).toBe('<strong>Lorem ipsum</strong>');
        done();
      });
    }

    render(
        <TestPopover
          onInserted={onInserted}
          renderContent={() => <TestComponent />}
        />,
    );
    clickPopover();
  });

  test('properly unmounts react component in popover', (done) => {
    render(
        <TestPopover
          onInserted={delayedClickPopover}
          onHidden={() => done()}
          renderContent={() => <TestComponent />}
        />,
    );
    clickPopover();
  });
});

describe('bootstrap apis', () => {
  test('component object exposed', (done) => {
    /**
     * A popover that will call the bootstrap show() method
     * immediately on mount.
     */
    function TestComponent(props) {
      const component = useRef();
      useEffect(() => component.current.show());
      return <TestPopover component={component} {...props} />;
    }

    /**
     * show.bs.popover event handler.
     */
    function onShow() {
      setTimeout(() => done(), 500);
    }

    render(<TestComponent onShow={onShow} />);
  });

  test('config prop passed to component', (done) => {
    /**
     * Confirm the placement of the popover is bottom.
     */
    function confirmPlacement() {
      const bottomPlacementPopover =
        document.querySelector('[data-popper-placement="bottom"]');
      expect(bottomPlacementPopover).not.toBeNull();
      setTimeout(() => done(), 500);
    }

    render(
        <TestPopover
          config={{placement: 'bottom'}}
          onShown={confirmPlacement}
        />,
    );
    clickPopover();
  });
});

describe('markup', () => {
  test('minimum markup renders as expected', () => {
    const {container} = render(<Popover />);
    const snapshots = {
      '5.2': `"<span></span>"`,
      '5.1': `"<span data-bs-original-title=\\"\\" title=\\"\\"></span>"`,
      '5.0': `"<span data-bs-original-title=\\"\\" title=\\"\\"></span>"`,
    };

    expect(container.innerHTML).toMatchInlineSnapshot(snapshots[majorMinor]);
  });

  test('places children in correct element', () => {
    const {container} = render(<Popover>child</Popover>);
    const snapshots = {
      '5.2': `"<span>child</span>"`,
      '5.1': `"<span data-bs-original-title=\\"\\" title=\\"\\">child</span>"`,
      '5.0': `"<span data-bs-original-title=\\"\\" title=\\"\\">child</span>"`,
    };

    expect(container.innerHTML).toMatchInlineSnapshot(snapshots[majorMinor]);
  });

  test('support additional attributes, place them correctly', () => {
    const {container} = render(<Popover data-lorem="ipsum" />);
    const snapshots = {
      '5.2': `"<span data-lorem=\\"ipsum\\"></span>"`,
      // eslint-disable-next-line max-len
      '5.1': `"<span data-lorem=\\"ipsum\\" data-bs-original-title=\\"\\" title=\\"\\"></span>"`,
      // eslint-disable-next-line max-len
      '5.0': `"<span data-lorem=\\"ipsum\\" data-bs-original-title=\\"\\" title=\\"\\"></span>"`,
    };

    expect(container.innerHTML).toMatchInlineSnapshot(snapshots[majorMinor]);
  });

  test('supports custom tagname by string', () => {
    const {container} = render(<Popover as="button" />);
    const snapshots = {
      '5.2': `"<button></button>"`,
      '5.1': `"<button data-bs-original-title=\\"\\" title=\\"\\"></button>"`,
      '5.0': `"<button data-bs-original-title=\\"\\" title=\\"\\"></button>"`,
    };

    expect(container.innerHTML).toMatchInlineSnapshot(snapshots[majorMinor]);
  });

  test('support custom tagname with components supporting ref forwards', () => {
    /**
     * A component that renders a link, and importantly supports ref forwarding.
     */
    const Link = forwardRef((props, ref) => {
      const {children, ...rest} = props;
      return <a ref={ref} {...rest}>{children}</a>;
    });

    Link.displayName = 'Link';

    Link.propTypes = {
      children: PropTypes.node,
    };

    const {container} = render(<Popover as={Link}>child</Popover>);
    const snapshots = {
      '5.2': `"<a>child</a>"`,
      '5.1': `"<a data-bs-original-title=\\"\\" title=\\"\\">child</a>"`,
      '5.0': `"<a data-bs-original-title=\\"\\" title=\\"\\">child</a>"`,
    };

    expect(container.innerHTML).toMatchInlineSnapshot(snapshots[majorMinor]);
  });
});
