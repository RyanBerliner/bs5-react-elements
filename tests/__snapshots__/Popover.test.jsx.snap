// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markup minimum markup renders as expected: 5.0 1`] = `"<span data-bs-original-title=\\"\\" title=\\"\\"></span>"`;

exports[`markup minimum markup renders as expected: 5.1 1`] = `"<span data-bs-original-title=\\"\\" title=\\"\\"></span>"`;

exports[`markup minimum markup renders as expected: 5.2 1`] = `"<span></span>"`;

exports[`markup places children in correct element: 5.0 1`] = `"<span data-bs-original-title=\\"\\" title=\\"\\">child</span>"`;

exports[`markup places children in correct element: 5.1 1`] = `"<span data-bs-original-title=\\"\\" title=\\"\\">child</span>"`;

exports[`markup places children in correct element: 5.2 1`] = `"<span>child</span>"`;

exports[`markup support additional attributes, place them correctly: 5.0 1`] = `"<span data-lorem=\\"ipsum\\" data-bs-original-title=\\"\\" title=\\"\\"></span>"`;

exports[`markup support additional attributes, place them correctly: 5.1 1`] = `"<span data-lorem=\\"ipsum\\" data-bs-original-title=\\"\\" title=\\"\\"></span>"`;

exports[`markup support additional attributes, place them correctly: 5.2 1`] = `"<span data-lorem=\\"ipsum\\"></span>"`;

exports[`markup support custom tagname with components supporting ref forwards: 5.0 1`] = `"<a data-bs-original-title=\\"\\" title=\\"\\">child</a>"`;

exports[`markup support custom tagname with components supporting ref forwards: 5.1 1`] = `"<a data-bs-original-title=\\"\\" title=\\"\\">child</a>"`;

exports[`markup support custom tagname with components supporting ref forwards: 5.2 1`] = `"<a>child</a>"`;

exports[`markup supports custom tagname by string: 5.0 1`] = `"<button data-bs-original-title=\\"\\" title=\\"\\"></button>"`;

exports[`markup supports custom tagname by string: 5.1 1`] = `"<button data-bs-original-title=\\"\\" title=\\"\\"></button>"`;

exports[`markup supports custom tagname by string: 5.2 1`] = `"<button></button>"`;
